File: MinMax/minmax.cpp
#include <iostream>
#include <tuple>
using namespace std;

tuple<int,int> minMax(int arr[], int i, int j, int max, int min) {
	int max1, min1;
	if(i == j) {
		max = arr[i];
		min = arr[i];
	}
	else if(i == j-1) {
		if(arr[i] < arr[j]) {
			max = arr[j];
			min = arr[i];
		}
		else {
			max = arr[i];
			min = arr[j];
		}
	}
	else {
		int mid = (i+j)/2;
		tie(min,max) = minMax(arr,i,mid,max,min);
		tie(min1,max1) = minMax(arr,mid+1,j,max1,min1);
		if(max < max1) {
			max = max1;
		}
		if(min > min1) {
			min = min1;
		}
	}
	return make_tuple(min,max);
}

int main() {
	int max,min,a,b;
	int n;
	cout << "Enter number of elements: ";
	cin >> n;
	int arr[n+1];
	for(int i = 1; i <= n; i++) {
		cout << "Element-" << i << ": ";
		cin >> arr[i];
	}
	tie(a,b) = minMax(arr,1,n,max,min);
	cout << "Minimum element: " << a << endl;
	cout << "Maximum element: " << b << endl;
}

File: MinMax/minmax.lua
function minmax(arr, i, j, max, min)
	local max1, min1
	
	if i == j then
		max, min = arr[i], arr[i]
	elseif i == j - 1 then
		if arr[i] < arr[j] then
			max, min = arr[j], arr[i]
		else
			max, min = arr[i], arr[j]
		end
	else
		local mid = math.floor((i+j)/2)
		min, max = minmax(arr, i, mid, max, min)
		min1, max1 = minmax(arr, mid+1, j, max1, min1)

		if max < max1 then
			max = max1
		end
		
		if min > min1 then
			min = min1
		end
	end
		
	return min, max
end

function main()
	local min, max
	io.write("Enter number of elements: ")
	local n = io.read("*n")
	arr = {}
	for i = 1, n do
		io.write("Enter an element: ")
		arr[i] = io.read("*n")
	end

	min, max = minmax(arr, 1, n, max, min)
	print("Maximum element: "..max)
	print("Minimum element: "..min)
end

main()
